File: frontend/optool/src/types/assets.ts
export type FieldPrimitive = string | number | boolean | null
export type FieldValue = FieldPrimitive | Record<string, unknown> | unknown[]
export type FieldsMap = Record<string, FieldValue>

export type ServerAsset = {
  id: string
  ip: string
  name: string
  assetId?: string | null
  assetNo?: string | null
  fields: FieldsMap
  createdAt?: string | null
  createdBy?: string | null
  updatedAt?: string | null
  updatedBy?: string | null
  version?: number | null
  isDeleted?: boolean | null
  deleteReason?: string | null
  deletedAt?: string | null
  deletedBy?: string | null
}

export type AssetAction = 'CREATE' | 'UPDATE' | 'DELETE'
export type HistoryDiff = { path: string; before: unknown; after: unknown }

export type AssetHistory = {
  id: string
  assetId: string
  action: AssetAction
  changedAt: string

  changedBy?: string | null
  createdBy?: string | null

  patch?: Record<string, unknown> | null
  diff?: HistoryDiff[] | null
}

export const EOS_STATUS_KEY = 'eos_action_status' as const
export const EOS_DATE_KEY = 'eos_date' as const

export const eosStatusOptions = [
  { label: '지원 기간 중', value: 'ACTIVE' },
  { label: 'EoS 지남', value: 'EOS' },
] as const

export type EosActionStatus = typeof eosStatusOptions[number]['value']

File: app/models/assets.py
# app/models/assets.py
from datetime import datetime
from typing import Any, Dict, Optional, Literal, List

from pydantic import BaseModel, Field

# history action
AssetAction = Literal["CREATE", "UPDATE", "DELETE", "PURGE"]


class ServerAssetBase(BaseModel):
    ip: str = Field(default='')
    name: str = Field(..., min_length=1)
    asset_id: Optional[str] = Field(default=None)  # 유형별 PK (사용자 입력, 고유)
    asset_no: Optional[str] = Field(default=None)  # 자산번호 (일반 필드, 중복 허용)
    fields: Dict[str, Any] = Field(default_factory=dict)


class ServerAssetCreate(ServerAssetBase):
    pass


class ServerAssetReplace(ServerAssetBase):
    pass


class ServerAssetPatch(BaseModel):
    ip: Optional[str] = None
    name: Optional[str] = None
    asset_id: Optional[str] = None
    asset_no: Optional[str] = None
    fields: Optional[Dict[str, Any]] = None
    version: Optional[int] = None  # optimistic lock (optional)


class ServerAssetDelete(BaseModel):
    reason: Optional[str] = None


class ServerAssetOut(BaseModel):
    id: str
    ip: str
    name: str
    asset_id: Optional[str] = None
    asset_no: Optional[str] = None
    fields: Dict[str, Any]

    created_at: Optional[datetime] = None
    created_by: Optional[str] = None
    updated_at: Optional[datetime] = None
    updated_by: Optional[str] = None
    version: Optional[int] = None
    is_deleted: Optional[bool] = None
    delete_reason: Optional[str] = None
    deleted_at: Optional[datetime] = None
    deleted_by: Optional[str] = None


class AssetHistoryOut(BaseModel):
    id: str
    asset_id: str
    action: AssetAction
    changed_at: datetime
    changed_by: str
    patch: Optional[Dict[str, Any]] = None
    diff: Optional[List[dict]] = None
